.travis.yml

```diff
@@ -1,41 +1,160 @@
+    # NOTE: Disable core optimizations, significantly speeds up the build
+    - disableOptimization=true
+    - buildStackProject=false
+    # NOTE: Modify a Haskell package to add shell completion scripts for the given executable produced by it. These completion scripts will be picked up automatically if the resulting derivation is installed, e.g. by `nix-env -i`.
+    # Invocation:
+    #   generateOptparseApplicativeCompletions command pkg
+    #
+    #   command: name of an executable
+    #       pkg: Haskell package that builds the executables
+    - generateOptparseApplicativeCompletion=false
+    # NOTE: Don't fail at configure time if there are multiple versions of the same package in the (recursive) dependencies of the package being built. Will delay failures, if any, to compile time.
+    - allowInconsistentDependencies=false
+    # NOTE: Log file to dump GHCJS build into
+    - ghcjsTmpLogFile='/tmp/ghcjsTmpLogFile.jog'
+    # NOTE: Length of the GHCJS log tail (<40000)
+    - ghcjsLogTailLength=10000
+  # NOTE: {os} x {jobs} - {exclude} = {build matrix}
+  jobs:
+    - GHCVERSION=ghc865
+    - GHCVERSION=ghc883
+    - GHCVERSION=ghc8101
+    - GHCVERSION=ghcjs
 
-  matrix:
-    - GHCVERSION=ghc865 STRICT=false TRACING=false
-#     - GHCVERSION=ghcjs
-# 
-# matrix:
-#   allow_failures:
-#     - env: GHCVERSION=ghcjs
+#  2020-05-25: NOTE: We still yet to build this in `master`
+jobs:
+  allow_failures:
+    - env: GHCVERSION=ghcjs
+    - env: GHCVERSION=ghc8101
+  #  2020-05-26: NOTE: Discard macOS GHCJS build + -> CI loop becomes ~30-40 minutes shorter
+  exclude:
+    - os: osx
+      env: GHCVERSION=ghcjs
+    - os: osx
+      env: GHCVERSION=ghc8101
+    - os: osx
+      env: GHCVERSION=ghc865
+    # NOTE: Discard Linux GHC 8.8.3 in favour of macOS GHC 8.8.3, Nixpkgs channel is the same and Nix package code is the same, since we test other Linux deployments and since 8.8.3 builds on macOS - we can be sure it builds inside Nix in Linux also. So lets optimize number of CI builds
+    - os: linux
+      env: GHCVERSION=ghc883
 
 before_script:
   - sudo mount -o remount,exec,size=4G,mode=755 /run/user || true
   - sudo mkdir -p /etc/nix
   - echo "trusted-users = root $USER" | sudo tee -a /etc/nix/nix.conf
+  # NOTE: macOS service restart
   - sudo launchctl kickstart -k system/org.nixos.nix-daemon || true
 
 script:
-  - nix-env -iA cachix -f https://github.com/NixOS/nixpkgs/tarball/db557aab7b690f5e0e3348459f2e4dc8fd0d9298
-  - cachix use hnix
-  - if [ ! "$CACHIX_SIGNING_KEY" = "" ]; then cachix push hnix --watch-store& fi
-  - if [ ! "$CACHIX_SIGNING_KEY" = "" ]; then ./build.sh | cachix push hnix; else ./build.sh; fi
+  #
+  #
+  # NOTE: Install Cachix client using Nix:
+  - nix-env -iA cachix -f https://cachix.org/api/v1/install
+  - cachix use "$name"
+  # NOTE: If key is set - use Cachix push, else - proceed without it
+  - if [ ! "$CACHIX_SIGNING_KEY" = "" ]; then cachix push "$name" --watch-store& fi
```

Comment:
Aside: Why do we `cachix push` before building? Isn't it the build product that we want to push?

Reply:
Well, then I first would need to explain how `Cachix` works.

And how it goes with the CI.

It is a literal Domen line from the current `master`.

The short answer is: the Cachix `watches the store changes`:
```
  -w,--watch-store         Run in daemon mode and push store paths as they are
                           added to /nix/store
```

Since it is forked into background thread - it able to asynchronously upload the Nix packages into the Cachix remote cache storage during the prolonged builds are are happening, so it does not wastes the time.

Reply:
Oh, that sounds awesome! Lots more to learn for me though! ;)